calico/felix/ipsets.py

```diff
@@ -351,61 +344,57 @@ def _finish_msg_batch(self, batch, results):
 
     def _sync_to_ipset(self):
         _log.debug("Setting ipset %s to %s", self.name, self.members)
-        fd, filename = tempfile.mkstemp(text=True)
-        f = os.fdopen(fd, "w")
-
-        if not self.set_exists:
-            # ipset does not exist, so just create it and put the data in it.
-            set_name = self.name
-            create = True
-            swap = False
-        elif not self.tmpset_exists:
-            # Set exists, but tmpset does not
-            set_name = self.tmpname
-            create = True
-            swap = True
-        else:
-            # Both set and tmpset exist
-            set_name = self.tmpname
-            create = False
-            swap = True
 
-        if create:
-            f.write("create %s hash:ip family %s\n" % (set_name, self.family))
-        else:
-            f.write("flush %s\n" % (set_name))
-
-        for member in self.members:
-            f.write("add %s %s\n" % (set_name, member))
-
-        if swap:
-            f.write("swap %s %s\n" % (self.name, self.tmpname))
-            f.write("destroy %s\n" % (self.tmpname))
-
-        f.close()
-
-        # Load that data.
-        futils.check_call(["ipset", "restore", "-file", filename])
-
-        # By the time we get here, the set exists, and the tmpset does not if
-        # we just destroyed it after a swap (it might still exist if it did and
-        # the main set did not when we started, unlikely though that seems!).
-        self.set_exists = True
-        if swap:
-            self.tmpset_exists = False
-
-        # Tidy up the tmp file.
-        os.remove(filename)
+        # We use ipset restore, which processes a batch of ipset updates.
+        # The only operation that we're sure is atomic is swapping two ipsets
+        # so we build up the complete set of members in a temporary ipset,
+        # swap it into place and then delete the old ipset.
+        create_cmd = "create %s hash:ip family %s --exist"
+        input_lines = [
+            # Ensure both the main set and the temporary set exist.
+            create_cmd % (self.name, self.family),
+            create_cmd % (self.tmpname, self.family),
+
+            # Flush the temporary set, this is a no-op unless we had a
+            # left-over temporary set before.
+            "flush %s" % self.tmpname,
+        ]
+        # Add all the members to the temporary set,
+        input_lines += ["add %s %s" % (self.tmpname, m) for m in self.members]
+        # Then, atomically swap the temporary set into place.
+        input_lines.append("swap %s %s" % (self.name, self.tmpname))
+        # Finally, delete the temporary set (which was the old active set).
+        input_lines.append("destroy %s" % self.tmpname)
+        # COMMIT tells ipset restore to actually execute the changes.
+        input_lines.append("COMMIT")
+
+        input_str = "\n".join(input_lines) + "\n"
+        cmd = ["ipset", "restore"]
+        process = subprocess.Popen(cmd,
```

Comment:
I don't like calling subprocess directly. We deliberately always do subprocess commands going through futils, and I think you should use futils.check_call (which does involve giving it a default stdin argument).